<a href="https://colab.research.google.com/github/ArmanKhan-git/Credit_Card_Fraud_Detection/blob/main/Credit_Card_Fraud.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

In [ ]:
df=pd.read_csv("creditcard.csv")
df.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62    0.0  
1  0.125895 -0.008983  0.014724    2.69    0.0  
2 -0.139097 -0.055353 -0.059752  378.66    0.0  
3 -0.221929  0.062723  0.061458  123.50    0.0  
4  0.502292  0.219422  0.215153   69.99    0.0  

[5 rows x 31 columns]

# Info about data

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 166510 entries, 0 to 166509
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    166510 non-null  float64
 1   V1      166510 non-null  float64
 2   V2      166510 non-null  float64
 3   V3      166510 non-null  float64
 4   V4      166510 non-null  float64
 5   V5      166510 non-null  float64
 6   V6      166510 non-null  float64
 7   V7      166510 non-null  float64
 8   V8      166509 non-null  float64
 9   V9      166509 non-null  float64
 10  V10     166509 non-null  float64
 11  V11     166509 non-null  float64
 12  V12     166509 non-null  float64
 13  V13     166509 non-null  float64
 14  V14     166509 non-null  float64
 15  V15     166509 non-null  float64
 16  V16     166509 non-null  float64
 17  V17     166509 non-null  float64
 18  V18     166509 non-null  float64
 19  V19     166509 non-null  float64
 20  V20     166509 non-null  float64
 21  V21     16

In [ ]:
df.describe()

Time             V1             V2             V3  \
count  166510.000000  166510.000000  166510.000000  166510.000000   
mean    59562.386956      -0.183809       0.047626       0.523413   
std     26520.811342       1.845073       1.607238       1.371207   
min         0.000000     -56.407510     -72.715728     -33.680984   
25%     40765.000000      -0.994264      -0.531828      -0.023844   
50%     59759.500000      -0.202150       0.115220       0.644815   
75%     77457.000000       1.179660       0.807645       1.309122   
max    118120.000000       2.439207      22.057729       9.382558   

                  V4             V5             V6             V7  \
count  166510.000000  166510.000000  166510.000000  166510.000000   
mean        0.125865      -0.188051       0.061410      -0.086538   
std         1.367585       1.334149       1.292984       1.205265   
min        -5.519697     -42.147898     -26.160506     -43.557242   
25%        -0.733079      -0.837511      -0.686988      -0.587720   
50%         0.132825      -0.235912      -0.199559      -0.035196   
75%         0.947324       0.357933       0.451325       0.456219   
max        16.875344      34.801666      22.529298      36.677268   

                  V8             V9  ...            V21            V22  \
count  166509.000000  166509.000000  ...  166509.000000  166509.000000   
mean        0.036431       0.020677  ...      -0.031555      -0.089674   
std         1.231459       1.155562  ...       0.743529       0.662630   
min       -73.216718     -13.434066  ...     -34.830382     -10.933144   
25%        -0.158796      -0.659710  ...      -0.231334      -0.548361   
50%         0.059746      -0.080024  ...      -0.056455      -0.071308   
75%         0.354005       0.646222  ...       0.124019       0.353865   
max        20.007208      15.594995  ...      27.202839      10.503090   

                 V23            V24            V25            V26  \
count  166509.000000  166509.000000  166509.000000  166509.000000   
mean       -0.023931       0.009833       0.097998       0.013392   
std         0.581129       0.598004       0.460916       0.491248   
min       -44.807735      -2.836627     -10.295397      -2.604551   
25%        -0.170534      -0.330922      -0.186486      -0.330421   
50%        -0.037522       0.060633       0.141450      -0.059333   
75%         0.095999       0.414848       0.402274       0.274142   
max        19.002942       4.022866       7.519589       3.517346   

                 V27            V28         Amount          Class  
count  166509.000000  166509.000000  166509.000000  166509.000000  
mean        0.002173       0.002487      86.912851       0.002150  
std         0.391714       0.308553     243.303381       0.046319  
min       -22.565679     -11.710896       0.000000       0.000000  
25%        -0.064629      -0.023004       5.380000       0.000000  
50%         0.009167       0.021554      21.490000       0.000000  
75%         0.089690       0.078277      76.150000       0.000000  
max        12.152401      33.847808   19656.530000       1.000000  

[8 rows x 31 columns]

In [ ]:
df=df.dropna(subset=['V23','V24','V25','V26','V27','V28','Amount','Class'])

In [ ]:
df.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

In [ ]:
df['Class'].value_counts()

Class
0.0    166151
1.0       358
Name: count, dtype: int64

# Data is highly imbalance

In [ ]:
# seperating the data into legit and fraud
legit=df[df['Class']==0]
fraud=df[df['Class']==1]

In [ ]:
# now statistical measures of the data
print(legit['Amount'].describe())
print(fraud['Amount'].describe())

count    166151.000000
mean         86.859640
std         243.333364
min           0.000000
25%           5.400000
50%          21.500000
75%          76.070000
max       19656.530000
Name: Amount, dtype: float64
count     358.000000
mean      111.608492
std       227.944683
min         0.000000
25%         1.000000
50%        10.685000
75%       101.500000
max      1809.680000
Name: Amount, dtype: float64


In [ ]:
df.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0.0    59569.056190 -0.171198  0.038284  0.541751  0.115616 -0.179010   
1.0    56303.567039 -6.025123  4.401259 -7.982396  4.882663 -4.383916   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0.0    0.064653 -0.071571  0.034840  0.026915  ...  0.031567 -0.033275   
1.0   -1.448126 -7.031758  0.774862 -2.874111  ...  0.413668  0.766540   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0.0   -0.089830 -0.023887  0.010016  0.098037  0.013342  0.001784  0.002396   
1.0   -0.017387 -0.044248 -0.075042  0.080196  0.036588  0.182343  0.044561   

           Amount  
Class              
0.0     86.859640  
1.0    111.608492  

[2 rows x 30 columns]

In [ ]:
# scaling amount and time
from sklearn.preprocessing import StandardScaler
scaler=StandardScaler()
cols=['Time','Amount']
df[cols]=scaler.fit_transform(df[cols])


# Now due to imbalance we will take a sample dataset but first we gotta split the dataset cause if i done balancing now then it wont reflect to the real world scenario so i gotta do it in train dataset only


In [ ]:
from sklearn.model_selection import train_test_split
X=df.drop(columns=['Class'])
y=df['Class']

In [ ]:
from imblearn.over_sampling import SMOTE
smote = SMOTE(random_state=42)
X_res, y_res = smote.fit_resample(X, y)


In [ ]:
y_res.value_counts()

Class
0.0    166151
1.0    166151
Name: count, dtype: int64

# Taking out the sample from this population

In [ ]:
from sklearn.model_selection import train_test_split

# Let's take 50k rows as an example
X_sample, _, y_sample, _ = train_test_split(
    X_res, y_res,
    train_size=50000,       # total rows in sample
    stratify=y_res,         # maintain same class ratio
    random_state=42
)



In [ ]:
y_sample.value_counts()

Class
1.0    25000
0.0    25000
Name: count, dtype: int64

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X_sample, y_sample, test_size=0.20,stratify=y_sample, random_state=42)

In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import accuracy_score, f1_score,precision_score,recall_score,roc_auc_score,confusion_matrix

In [ ]:
models = {
    'Logistic Regression': LogisticRegression(),
    'SVM': SVC(),
    'KNN': KNeighborsClassifier(),
    'Decision Tree': DecisionTreeClassifier(),
    'Naive Bayes': GaussianNB()
}

results=[]

In [ ]:
for name,model in models.items():
  model.fit(X_train,y_train)
  y_pred=model.predict(X_test)
  results.append({
      'model':name,
      'accuracy': accuracy_score(y_test, y_pred),
      'f1': f1_score(y_test, y_pred),
      'Precision':precision_score(y_test,y_pred),
      'Recall':recall_score(y_test,y_pred),
      'ROC-AUC':roc_auc_score(y_test,y_pred)
  })

In [ ]:
results

[{'model': 'Logistic Regression',
  'accuracy': 0.9556,
  'f1': 0.9545454545454546,
  'Precision': 0.977768456375839,
  'Recall': 0.9324,
  'ROC-AUC': np.float64(0.9555999999999999)},
 {'model': 'SVM',
  'accuracy': 0.9649,
  'f1': 0.964209238299174,
  'Precision': 0.983565633451217,
  'Recall': 0.9456,
  'ROC-AUC': np.float64(0.9648999999999999)},
 {'model': 'KNN',
  'accuracy': 0.9945,
  'f1': 0.9945300845350572,
  'Precision': 0.9891196834817013,
  'Recall': 1.0,
  'ROC-AUC': np.float64(0.9944999999999999)},
 {'model': 'Decision Tree',
  'accuracy': 0.9905,
  'f1': 0.9905217998603213,
  'Precision': 0.9882540314553055,
  'Recall': 0.9928,
  'ROC-AUC': np.float64(0.9904999999999998)},
 {'model': 'Naive Bayes',
  'accuracy': 0.9239,
  'f1': 0.9194112040665043,
  'Precision': 0.9770425388251182,
  'Recall': 0.8682,
  'ROC-AUC': np.float64(0.9239)}]

In [ ]:
results_train=[]
for name,model in models.items():
  results_train.append({
      'model':name,
      'score': model.score(X_train,y_train)
  })

In [ ]:
results_train

[{'model': 'Logistic Regression', 'score': 0.954675},
 {'model': 'SVM', 'score': 0.9643},
 {'model': 'KNN', 'score': 0.9962},
 {'model': 'Decision Tree', 'score': 1.0},
 {'model': 'Naive Bayes', 'score': 0.9245}]

🔹 Observations

KNN

Accuracy: 0.9942

F1: 0.9942

Precision: 0.9885

Recall: 1.0 → catches all fraud cases

ROC-AUC: 0.9942

✅ Clearly your best model. F1, Precision, and Recall are all excellent, meaning it’s almost perfect on both classes.

Decision Tree

Accuracy: 0.987

F1: 0.987

Precision: 0.984, Recall: 0.99

Very strong, slightly worse than KNN. Good backup.

SVM

Accuracy: 0.955

F1: 0.954

Precision: 0.983, Recall: 0.9266

Decent, but recall is lower than KNN → misses some frauds.

Logistic Regression

Accuracy: 0.946

F1: 0.9447

Precision: 0.973, Recall: 0.918

Lower recall than SVM & KNN → less suitable for fraud detection.

Naive Bayes

Accuracy: 0.9138

F1: 0.9082

Recall: 0.8532 → misses more frauds

Not suitable here

In [ ]:
from sklearn.model_selection import GridSearchCV

In [ ]:
classifier_knn = GridSearchCV(
    KNeighborsClassifier(),
    param_grid={
        'n_neighbors': [3, 5, 7, 9, 11],
        'weights': ['uniform', 'distance'],
        'metric': ['minkowski', 'euclidean', 'manhattan']
    },
    cv=5,
    return_train_score=False
)

In [ ]:
classifier_knn.fit(X_train,y_train)

GridSearchCV(cv=5, estimator=KNeighborsClassifier(),
             param_grid={'metric': ['minkowski', 'euclidean', 'manhattan'],
                         'n_neighbors': [3, 5, 7, 9, 11],
                         'weights': ['uniform', 'distance']})

In [ ]:
y_pred_best = classifier_knn.best_estimator_.predict(X_test)
y_prob_best = classifier_knn.best_estimator_.predict_proba(X_test)[:, 1]

In [ ]:
acc=accuracy_score(y_test,y_pred_best)
f1=f1_score(y_test,y_pred_best)
recall=recall_score(y_test,y_pred_best)
precision=precision_score(y_test,y_pred_best)
Roc_score=roc_auc_score(y_test,y_prob_best)

In [ ]:
print(acc)
print(f1)
print(recall)
print(precision)
print(Roc_score)

0.997
0.9970089730807578
1.0
0.9940357852882704
0.9984997


# So gridsearchcv do boost up our model now lets see confusuion metric of knn and knn_best

In [ ]:
knn=KNeighborsClassifier()
knn.fit(X_train,y_train)
y_pred_knn=knn.predict(X_test)

In [ ]:
knn_metrix=confusion_matrix(y_test,y_pred_knn)
knn_best=confusion_matrix(y_test,y_pred_best)

In [ ]:
print(knn_metrix)
print(knn_best)

[[4945   55]
 [   0 5000]]
[[4970   30]
 [   0 5000]]


# In knn:
[[4945   55]
 [   0 5000]]
True Negatives (TN): 4945 → non-fraud correctly predicted

False Positives (FP): 55 → non-fraud predicted as fraud

False Negatives (FN): 0 → fraud missed = none ✅

True Positives (TP): 5000 → fraud correctly predicted

# In knn_best:
[[4970   30]
 [   0 5000]]
True Negatives (TN): 4970 → more non-fraud correctly predicted

False Positives (FP): 30 → fewer false alarms than normal KNN ✅

False Negatives (FN): 0 → still no fraud missed

True Positives (TP): 5000 → all fraud detected

✅ Precision improved (less FP), recall remains 1.0 → F1 also slightly improved.

# We got the best model now lets see if random forest does any better

In [ ]:
from sklearn.ensemble import BaggingClassifier,RandomForestClassifier

In [ ]:
model_rf=RandomForestClassifier(
    n_estimators=100,  # number of tree
    max_depth=None,  # lets tree grows fully
    random_state=42,
    max_features='sqrt' # matlab jitne columns hai unka square root(4 Cols hai to 2 lelega) nikalke utne columns mese select krega root node
)

In [ ]:
model_rf.fit(X_train,y_train)

RandomForestClassifier(random_state=42)

In [ ]:
y_pred_rf=model_rf.predict(X_test)
y_prob_rf = model_rf.predict_proba(X_test)[:, 1]
acc=accuracy_score(y_test,y_pred_rf)
f1=f1_score(y_test,y_pred_rf)
recall=recall_score(y_test,y_pred_rf)
precision=precision_score(y_test,y_pred_best)
Roc_score=roc_auc_score(y_test,y_prob_rf)

In [ ]:
print(acc)
print(f1)
print(recall)
print(precision)
print(Roc_score)

0.997
0.9970089730807578
1.0
0.9940357852882704
0.9984997


In [ ]:
rf_matrix=confusion_matrix(y_test,y_pred_rf)

In [ ]:
print(knn_best)
print(rf_matrix)

[[4970   30]
 [   0 5000]]
[[4996    4]
 [   7 4993]]


# knn_best
# [[4970  , 30]
# [   0 , 5000]]
True Negatives (TN): 4970 → correctly predicted non-fraud

False Positives (FP): 30 → non-fraud misclassified as fraud

False Negatives (FN): 0 → all fraud detected ✅

True Positives (TP): 5000 → fraud correctly predicted

Observations:

Recall = 1.0 (no fraud missed)

Precision slightly < 1 due to 30 FPs

# randomforest
# [[4996  ,  4]
#  [   7 ,  4993]]
True Negatives (TN): 4996 → more non-fraud correctly predicted than KNN

False Positives (FP): 4 → fewer false alarms ✅

False Negatives (FN): 7 → 7 fraud cases missed

True Positives (TP): 4993 → slightly lower than KNN

Observations:

Precision is very high (FP only 4)

Recall slightly lower (FN = 7) → missed a few fraud cases

# KNN sare fraud catch krlera hai lekin RF nahi
# aise hi RF less false alarm dera hai lekin KNN zada dera hai to real world scenario we want to minimize the fraud so i m gonna use knn_best or classifier_knn model for this dataset

# So now me knn trai krunga smote data me oor test krunga real dataset pe


In [ ]:
# train_test_split on smote data
X_train_res, X_test_res, y_train_res, y_test_res = train_test_split(X_res, y_res, test_size=0.20,stratify=y_res, random_state=42)
knn=KNeighborsClassifier()
knn.fit(X_train_res,y_train_res)
y_pred_knn_smote=knn.predict(X_test_res)

In [ ]:
X_train_org, X_test_org, y_train_org, y_test_org = train_test_split(X, y, test_size=0.20, random_state=42)

In [ ]:
y_pred_org=knn.predict(X_test_org)
y_prob_org = knn.predict_proba(X_test_org)[:, 1]

In [ ]:
acc=accuracy_score(y_test_res,y_pred_knn_smote)
f1=f1_score(y_test_res,y_pred_knn_smote)
recall=recall_score(y_test_res,y_pred_knn_smote)
precision=precision_score(y_test_res,y_pred_knn_smote)
Roc_score=roc_auc_score(y_test_res,y_pred_knn_smote)

In [ ]:
print(acc)
print(f1)
print(recall)
print(precision)
print(Roc_score)

0.9989166578895894
0.9989178139842482
1.0
0.9978379676896283
0.9989166741897626


In [ ]:
acc=accuracy_score(y_test_org,y_pred_org)
f1=f1_score(y_test_org,y_pred_org)
recall=recall_score(y_test_org,y_pred_org)
precision=precision_score(y_test_org,y_pred_org)
Roc_score=roc_auc_score(y_test_org,y_prob_org)

In [ ]:
print(acc)
print(f1)
print(recall)
print(precision)
print(Roc_score)

0.9982583628610894
0.7040816326530612
1.0
0.5433070866141733
0.9999548641410646


In [ ]:
model_rf=RandomForestClassifier()
model_rf.fit(X_train_res,y_train_res)
y_pred_rf_smote=model_rf.predict(X_test_res)


In [ ]:
y_pred_org_rf=model_rf.predict(X_test_org)
y_prob_org_rf = model_rf.predict_proba(X_test_org)[:, 1]

In [ ]:
acc=accuracy_score(y_test_res,y_pred_rf_smote)
f1=f1_score(y_test_res,y_pred_rf_smote)
recall=recall_score(y_test_res,y_pred_rf_smote)
precision=precision_score(y_test_res,y_pred_rf_smote)
Roc_score=roc_auc_score(y_test_res,y_pred_rf_smote)

In [ ]:
print(acc)
print(f1)
print(recall)
print(precision)
print(Roc_score)

0.9999247679089993
0.9999247724366208
1.0
0.9998495561907628
0.9999247690409557


In [ ]:
acc=accuracy_score(y_test_org,y_pred_org)
f1=f1_score(y_test_org,y_pred_org_rf)
recall=recall_score(y_test_org,y_pred_org_rf)
precision=precision_score(y_test_org,y_pred_org_rf)
Roc_score=roc_auc_score(y_test_org,y_prob_org_rf)

In [ ]:
print(acc)
print(f1)
print(recall)
print(precision)
print(Roc_score)

0.9982583628610894
0.9928057553956835
1.0
0.9857142857142858
1.0


# it seems that randomforest is doing a great job rather than knn lets see the confusion matrix on original dataset

In [ ]:
knn_matrix_org=confusion_matrix(y_test_org,y_pred_org)
rf_matrix_org=confusion_matrix(y_test_org,y_pred_org_rf)

In [ ]:
print(knn_matrix_org)
print(rf_matrix_org)

[[33175    58]
 [    0    69]]
[[33232     1]
 [    0    69]]


# knn on orignal dataset
#[[33175  ,  58]
# [    0   , 69]]
True Negatives (TN) = 33175 → correctly predicted non-frauds

False Positives (FP) = 58 → non-frauds incorrectly predicted as fraud

False Negatives (FN) = 0 → no fraud missed ✅

True Positives (TP) = 69 → correctly detected frauds ✅

✅ Interpretation:

Recall = 1.0 → all frauds detected

Precision = 69 / (69 + 58) ≈ 0.543 → moderate because of false positives

# RandomForest
# [[33232  ,   1]
# [    0  ,  69]]
TN = 33232 → correctly predicted non-frauds

FP = 1 → almost no false positives

FN = 0 → all frauds detected ✅

TP = 69 → correctly detected frauds ✅

✅ Interpretation:

Recall = 1.0 → all frauds detected

Precision = 69 / (69 + 1) ≈ 0.9857 → excellent, very few false positives

Clearly superior to KNN in real-world performance



# So RandomForest is doing best on the original dataset

# Now lets extract the scaler rf model and columns

In [ ]:
import joblib
joblib.dump(model_rf,'RandomForest_model.pkl')
joblib.dump(scaler,'scaler.pkl')
columns=df.columns.to_list()
joblib.dump(columns,'columns.pkl')


['columns.pkl']